osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -0,0 +1,168 @@
+        this.encryptionProviderType = encryptionProviderType;
+        this.secretRepository = secretRepository;
+        this.keyRepository = keyRepository;
+    }
+
+    public void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) throws Exception {
+        //@formatter:off
+        secrets.stream()
+                .map(t -> this.validateSecret(deviceIdentification, t))
+                .map(s -> this.createEncrypted(deviceIdentification, s, this.getKey()))
+                .forEach(this.secretRepository::save);
+        //@formatter:on
+    }
+
+    private DbEncryptionKeyReference getKey() {
+        final Date now = new Date();
+        final Page<DbEncryptionKeyReference> keyRefsPage = this.keyRepository.findByTypeAndValid(now,
+                this.encryptionProviderType, Pageable.unpaged());
+        if (keyRefsPage.getSize() > 1) {
+            throw new IllegalStateException("Multiple encryption keys found that are valid at " + now);
+        }
+        return keyRefsPage.stream().findFirst().orElseThrow(
+                () -> new NoSuchElementException("No encryption key found that is valid at " + now));
+    }
+
+    private TypedSecret validateSecret(final String deviceIdentification, final TypedSecret secret) {
+        if (secret.getSecret() == null) {
+            throw new IllegalArgumentException("No secret string set");
+        } else if (secret.getSecretType() == null) {
+            throw new IllegalArgumentException("No secret type set");
+        } else if (this.isIdenticalToCurrent(deviceIdentification, secret)) {
+            throw new IllegalArgumentException(
+                    String.format("Secret is identical to current secret (%s, %s)", deviceIdentification,
+                            secret.getSecretType().name()));
+        }
+        return secret;
+    }
+
+    private boolean isIdenticalToCurrent(final String deviceIdentification, final TypedSecret secret) {
+        try {
+            final TypedSecret current = this.retrieveSecret(deviceIdentification, secret.getSecretType());
+            return current.getSecret().equals(secret.getSecret());
+        } catch (final NoSuchElementException nsee) {
+            //there is no current secret
+            return false;
+        }
+    }
+
+    private DbEncryptedSecret createEncrypted(final String deviceIdentification, final TypedSecret typedSecret,
+            final DbEncryptionKeyReference keyReference) {
+        final String secretString = typedSecret.getSecret();
+        final byte[] secretBytes = HexUtils.fromHexString(secretString);
+        final Secret secret = new Secret(secretBytes);
+        try {
+            final EncryptedSecret encryptedSecret = this.encryptionDelegate.encrypt(
+                    keyReference.getEncryptionProviderType(), secret, keyReference.getReference());
+            final DbEncryptedSecret dbEncryptedSecret = new DbEncryptedSecret();
```

Comment:
I would rewrite this into a constructor to make data immutable.

also it is much less lines. just return new DbEncryptedSecret(value1, value2) etc